bundles/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/internal/handler/NibeHeatPumpHandler.java

```diff
@@ -411,29 +422,67 @@ private void parseWriteEnabledItems() throws IllegalArgumentException {
         logger.debug("Enabled registers for write commands: {}", itemsToEnableWrite);
     }
 
-    private State convertNibeValueToState(NibeDataType dataType, double value, String acceptedItemType) {
+    private State convertNibeValueToState(VariableInformation variableInfo, int value, String acceptedItemType) {
         State state = UnDefType.UNDEF;
 
+        NibeDataType dataType = variableInfo.dataType;
+        int decimals = (int)Math.log10(variableInfo.factor);
+        byte[] bytes;
+        int x;
+        switch (dataType) {
```

Comment:
Your conversions seems rather complex to me, especially those signed values :)

Should following do the same (I didn't do any tests):

long v;

switch (dataType) {
	case U8:
		v = (byte) value & 0xFF;
		break;
	case U16:
		v = (short) value & 0xFFFF;
		break;
	case U32:
		v = value & 0xFFFFFFFFL;
		break;
	case S8:
		v = (byte) value;
		break;
	case S16:
		v = (short) value;
		break;
	case S32:
		v = value;
		break;
}

BigDecimal converted = new BigDecimal(v).movePointLeft(decimals).setScale(decimals, RoundingMode.HALF_EVEN);

Reply:
That will be too simple that everybody would understand. 😄 

Found another way. Much simpler.